### ISSUE TITLE
version-range: Support multiple different MSRVs in a workspace

### ISSUE THREAD

**Issue by taiki-e** [taiki-e/cargo-hack#199]
To implement this, we need to make sure that every crate in which the command is actually run has rust-version field, if there are multiple different MSRVs in a workspace.

https://github.com/taiki-e/cargo-hack/blob/1057aec30d897460a4379bd7dcea7b268e93c475/src/rustup.rs#L43-L61

---

A concrete example of such a workspace is futures-rs:
- futures-core: 1.36 https://github.com/rust-lang/futures-rs/blob/6dc49d003cc216cc7321041dbea7e12f5fc59a1a/futures-core/Cargo.toml#L5
- futures: 1.56 https://github.com/rust-lang/futures-rs/blob/6dc49d003cc216cc7321041dbea7e12f5fc59a1a/futures/Cargo.toml#L5

**Comment by epage** [comment]
If you want, I can take this on after #196 is fixed.  They'll conflict a lot with each other and we would need this if we are to use this in cargo.

**Comment by taiki-e** [comment]
That would be great!

**Comment by epage** [comment]
The most basic implementation is "pick the highest version in the workspace" (ie what will work for everything).  What cargo needs is an MSRV per package.  I can think of one of two ways to implement this and wanted to get your thoughts
- Defer evaluating `MaybeVersion::Msrv` from #202 until we get to package processing
- Instantiate a range per package

Any preferences?

**Comment by taiki-e** [comment]
> - Defer evaluating MaybeVersion::Msrv from feat: Add --rust-version flag  #202 until we get to package processing
> - Instantiate a range per package


I was originally trying to implement this with the former approach (evaluating MSRV after `determine_package_list`, which creates a list of packages to actually run command on). The patch can be found [here](https://github.com/taiki-e/cargo-hack/commit/1d647e877e030d5e56e7202187e23ed5ad35d894) (feel free to copy it if needed).

That patch is incomplete because there are the following two main problems, but I'm still thinking it is reasonable to take that approach.

1\. [`--version-step` is inaccurate](https://github.com/taiki-e/cargo-hack/commit/1d647e877e030d5e56e7202187e23ed5ad35d894#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR75-R81).
2\. [Build number counter is inaccurate](https://github.com/taiki-e/cargo-hack/commit/1d647e877e030d5e56e7202187e23ed5ad35d894#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR109-R110).

I think 1\. is a problem that does not exist in the "Instantiate a range per package" approach, but there is a problem that the "Instantiate a range per package" approach may increase the number of rust versions to be downloaded and compiled (e.g., if step is 2, crate `a`'s  MSRV is 63, crate `b`'s MSRV is 64, the range is [63, 65, 67, ...] for crate `a` and [64, 66, 68, ...] for crate `b`).

Personally, I think that if there is a crate with an MSRV greater than the lowest one, it is okay to just add that version to the range (e.g., if step is 2, crate `a`'s  MSRV is 63, crate `b`'s MSRV is 64, the range is [63, *64*, 65, 67, ..]).

Alternatively, it is also okay to implement the minimum ones and postpone `--version-step` support.

I think 2\. can be addressed by evaluating MSRV (at least partially) inside `determine_package_list`.
